_posts/2021-03-10-dask_memory_usage.md

```diff
@@ -0,0 +1,261 @@
+
+But in order to provision hardware appropriately, I needed to know how much peak memory each task was using.
+
+## Introducing `dask-memusage`
+
+`dask-memusage` is a tool for measuring peak memory usage for each task in the Dask execution graph.
+
+* Per *task* because Dask executes code as a graph of tasks, and the graph determines how much parallelism can be used.
+* *Peak* memory is important, because that is the bottleneck.
+  It doesn't matter if average memory usage per task is 4GB, if two tasks need 12GB each at the same time your computer needs 24GB of RAM to be available.
+
+The interaction of the task graph and parallelism best illustrated with an example.
+Let's consider some trivial code that squares some numbers and then sums them:
+
+```python
+from dask.bag import from_sequence
+
+def square(x):
+    return x * x
+
+result = from_sequence(range(5)).map(square).sum()
+```
+
+We can visualize the task graph, the series of steps that Dask will try to execute:
+
+```python
+result.visualize(filename="example1.png", rankdir="TD")
+```
+
+Here's what it looks like:
+
+<img src="/images/dask_memusage/example1.png"  style="width: 75%; margin: 2em;">
+
+As we'd expect, squaring can happen in parallel, but summing the resulting values eventually devolves to a single action without parallelism because eventually we're combining down to a single result.
+
+For memory use purposes, in this particular example the peak memory usage would worst case be the peak memory usage of `square()` multiplied by the level of parallelism.
+So in the highly unlikely event that squaring a number takes 1GB of RAM, we'd need 10GB of RAM if we ran 10 `square()` calls in parallel.
+
+# Using `dask-memusage`
+
+Let's try a slightly more realistic example: we're going to count the occurrence of words in some text files, and then report the top-10 most common words in each file.
+You can imagine we might also combine the data later on, but we won't bother with that in this simple example.
+
+```python
+import sys
+import gc
+from time import sleep
+from pathlib import Path
+from dask.bag import from_sequence
+from collections import Counter
+from dask.distributed import Client, LocalCluster
+import dask_memusage
+
+
+def calculate_top_10(file_path: Path):
+    gc.collect()  # See notes below
+
+    # Load the file
+    with open(file_path) as f:
+        data = f.read()
+
+    # Count the words
+    counts = Counter()
+    for word in data.split():
+        counts[word.strip(".,'\"").lower()] += 1
+
+    # Choose the top 10:
+    by_count = sorted(counts.items(), key=lambda x: x[1])
+    sleep(0.1)  # See notes below
+    return (file_path.name, by_count[-10:])
+
+
+def main(directory):
+    # Setup the calculation:
+
+    # Create a 4-process cluster (running locally). Note only one thread
+    # per-worker; this is important or you won't get correct results.
```

Comment:
This might be worth a little more elaboration, this sort of thing often trips people up.

Reply:
Ah, it looks like this is mentioned further down "Finally, because polling is per-process, you can't run multiple threads per worker, otherwise you'll get results that combine memory usage of multiple tasks.". Can we add that info in-line here too?